pyomo/repn/plugins/nl_writer.py

```diff
@@ -2328,23 +2305,46 @@ def _before_linear(visitor, child):
     linear = {}
     for arg in child.args:
         if arg.__class__ is MonomialTermExpression:
-            c, v = arg.args
-            if c.__class__ not in native_types:
-                c = visitor._eval_expr(c)
-            if v.fixed:
-                const += c * v.value
-            elif c:
-                _id = id(v)
-                if _id not in var_map:
-                    var_map[_id] = v
-                if _id in linear:
-                    linear[_id] += c
-                else:
-                    linear[_id] = c
+            arg1, arg2 = arg._args_
+            if arg1.__class__ not in native_types:
+                try:
+                    arg1 = visitor._eval_expr(arg1)
+                except (ValueError, ArithmeticError):
+                    return True, None
+
+            # Trap multiplication by 0 and nan.
+            if not arg1:
+                if arg2.fixed:
+                    arg2 = visitor._eval_fixed(arg2)
+                    if arg2 != arg2:
+                        deprecation_warning(
+                            f"Encountered {arg1}*{str(arg2.value)} in expression "
+                            "tree.  Mapping the NaN result to 0 for compatibility "
+                            "with the nl_v1 writer.  In the future, this NaN "
+                            "will be preserved/emitted to comply with IEEE-754.",
+                            version='6.6.0',
+                        )
+                continue
+
+            _id = id(arg2)
+            if _id not in var_map:
+                if arg2.fixed:
+                    const += arg1 * visitor._eval_fixed(arg2)
+                    continue
+                var_map[_id] = arg2
+                linear[_id] = arg1
+            elif _id in linear:
+                linear[_id] += arg1
+            else:
+                linear[_id] = arg1
         elif arg.__class__ in native_types:
             const += arg
         else:
-            const += visitor._eval_expr(arg)
+            try:
+                const += visitor._eval_expr(arg)
+            except (ValueError, ArithmeticError):
+                return True, None
```

Comment:
What is the implication of catching these errors and moving on here (or generally in this PR)? What happens when `True, None` is returned here? In the example in #2906, the nl file is produced with `nan` as a constant in it. Surprisingly, sending this nl file to Ipopt solves with `Optimal solution found`. If I edit the model to have an objective, e.g.
```python
import pyomo.environ as pyo 
m = pyo.ConcreteModel()
m.x = pyo.Var([1, 2, 3, 4]) 
m.x[1].fix()
m.x[2] = 1 
m.x[3] = 2 
m.x[4] = 3 
m.con = pyo.Constraint(expr= m.x[1] - ((m.x[2]*m.x[3])*m.x[4]) == 0)
m.obj = pyo.Objective(expr=m.x[1]**2 + m.x[2]**2 + m.x[3]**2 + m.x[4]**2)
m.write("file.nl")

solver = pyo.SolverFactory("ipopt")
solver.solve(m, tee=True)
```
then Ipopt errors with its "Invalid number" exception, which is what I would expect from NaN in the nl file. But I'm not sure whether this is better for the user than an error trying to evaluate or parse an expression. Should this behavior be controlled by `pyomo.repn.util.HALT_ON_EVALUATION_ERROR`?

Reply:
This is a somewhat rambling review comment. I'm mostly wondering if invalid numbers were always intended to propagate to the nl file.

Reply:
I tried this with the `nl_v1` writer, and the exact error we get seems to depend on how `None` is encountered. It could be an error like:
```
ValueError: No value for uninitialized NumericValue object x[1]
```
or an error like:
```
TypeError: unsupported operand type(s) for *: 'int' and 'NoneType'
```

Reply:
Some answers:
- returning `(True, None)` from a `_before_*` function tells the walker that it needs to descend into that expression.  So the idea is that if evaluating an expression led to an error / exception, then let this walker descend into it to try and locate / sort out exactly what the problem is.  When the first return value is `True`, the second return value is ignored.
- The goal here is to never emit an invalid number to the NL file, as those generally lead to parse errors - not Ipopt evaluation errors - and parse errors are really hard for a user to track down.  Now, invalid numbers (that are not representable in an NL file) generate errors within Pyomo at write time.
- Resolving the ValueError/TypeError inconsistency is one of the big goals of this refactor.

Reply:
> * returning `(True, None)` from a `_before_*` function tells the walker that it needs to descend into that expression.  So the idea is that if evaluating an expression led to an error / exception, then let this walker descend into it to try and locate / sort out exactly what the problem is.

How does the walker know to raise the error? And where does it get raised now?